### ISSUE TITLE
LDAP Auth

### ISSUE THREAD

**Issue by Legolars99** [causefx/Organizr#594]
Hi,

I'm having a few issues setting up the LDAP authentication. 
My config is the following:

LDAP Hostname: `ldap://[IP]:389`
Domain: `OU=Users,OU=Accounts,DC=DOMAIN,DC=net`

When trying to log in in I get the follwing errors in the log:

> USERNAME tried to sign-in with the wrong password

>LDPA could not authenticate

I'm using the correct password, it's working for every other AD login. 

In addition to that I'm getting the follwing error at the top of the page: 
> Warning: ldap_bind(): Unable to bind to server: Invalid credentials in /config/www/Dashboard/functions.php on line 37

Any advice?

**Comment by Cerothen** [comment]
Is your LDAP usernames case sensitive? I am pretty sure that it forces the username to lower right now.

**Comment by joshuaavalon** [comment]
I am face the same problem

**Comment by Cerothen** [comment]
Have you tried entering just the domain? eg not the whole DN.

Basically it just prepends the domain to the username and send the bind request to the indicated LDAP server.

**Comment by joshuaavalon** [comment]
@Cerothen
Here is my setup: [lsiocommunity/organizr](https://hub.docker.com/r/lsiocommunity/organizr/) for Organizr and [sharaku/docker-ldap](https://hub.docker.com/r/sharaku/ldap/) for LDAP. I tested LDAP with LDAPExplorerTool 2 for my another computer.

Here is the LDAP test user.
![](https://user-images.githubusercontent.com/7152420/30788756-c2ee3aca-a1d2-11e7-848c-3d5ca5b5a493.png)

Here is what I tried.
```
LDAP: ldap://192.168.1.13:389
Domain: OU=People,DC=example,DC=com
```
Login with ttest / test. It failed. Nginx error log shows `Unable to bind to server: Invalid DN syntax`.
The result is the same for using `example.com` and `example` for domain.

Then, I investigated the error and tried to fix it in `functions.php`.

```php
$ldap = ldap_connect('192.168.1.13','389');
if ($bind = ldap_bind($ldap, 'cn=test test,ou=People,dc=example,dc=com', $password))
```

I hard code the LDAP DN and the error message changed to `Unable to bind to server: Protocol error`. Then, I change the LDAP  version to 3.

```php
$ldap = ldap_connect('192.168.1.13','389');
ldap_set_option($ldap,LDAP_OPT_PROTOCOL_VERSION,3);
if ($bind = ldap_bind($ldap, 'cn=test test,ou=People,dc=example,dc=com', $password))
```

The new error message changed to `Unable to bind to server: Server is unwilling to perform` but I am able to login.

**Comment by Cerothen** [comment]
What happens if you change out ```'cn=test test,ou=People,dc=example,dc=com'``` for ```example\test test```?

**Comment by joshuaavalon** [comment]
@Cerothen `Unable to bind to server: Invalid DN syntax`. Also, Organizr seems to block `\` with `LOG IN ERROR: USER NAME DID NOT PASS VALIDATION`.

**Comment by joshuaavalon** [comment]
I would like to purpose fix on LDAP. But I don't know how the setting and UI should work on Organizr, so I could not do a pull request.
I  would like to explain the idea behind the fixes and let you guys implpement it. These fixes should be able to make the LDAP works. At least, you can make the new default value to not break any working system.

`functions.php:36`
**Original**
```php
$ldap = ldap_connect(implode(' ',$ldapServers));
if ($bind = ldap_bind($ldap, AUTHBACKENDDOMAIN.'\\'.$username, $password)) {
	writeLog("success", "LDAP authentication success"); 
	return true;
} else {
	writeLog("error", "LDPA could not authenticate"); 
	return false;
}
writeLog("error", "LDPA could not authenticate");      
return false;
```

**Fixes**
```php
$ldap = ldap_connect(implode(' ',$ldapServers));

// Active Directory requires both set option to work, openLDAP seems to work on it too
// Or may be provides two number input field for user to input: 
// "LDAP Protocol Version" & "LDAP Referrals" which default to 3 and 0
// Or disable them if the value < 0
ldap_set_option($ldap, LDAP_OPT_PROTOCOL_VERSION, 3);
ldap_set_option($ldap, LDAP_OPT_REFERRALS, 0);

// Add a field for user to input distinguished name format.
// LDAP seems to support different format depends on which LDAP backend is used.
// User should input like "EXAMPLE\%s", or "uid=%s,ou=People,dc=example,dc=com"
// The value should be in $dn_format then format with the $username.
// Default to  "AUTHBACKENDDOMAIN.'\\'.$username"
$bind = ldap_bind($ldap, sprinf($dn_format, $username), $password)

// Move the about statment away from if remove "Unable to bind to server: Server is unwilling to perform" error.
// I don't know why.
if ($bind) {
	writeLog("success", "LDAP authentication success"); 
	return true;
} else {
	writeLog("error", "LDPA could not authenticate"); 
	return false;
}
writeLog("error", "LDPA could not authenticate");      
return false;
```

**Comment by causefx** [comment]
@Cerothen whatcha think?

**Comment by causefx** [comment]
@joshuaavalon are you available?

**Comment by joshuaavalon** [comment]
@causefx 
There is still a few problem like: #667 

Advance settings page
`Basi Domain to use for LDAP` -> `Basic Domain to use for LDAP`
`Advanced Domain to use for LDAP i.e. uid=s,ou=People,dc=example,dc=com` -> `Advanced Domain to use for LDAP i.e. uid=%s,ou=People,dc=example,dc=com`

I tried out Synology Directory Server, it still throw `Unable to bind to server: Server is unwilling to perform" error`

https://github.com/causefx/Organizr/blob/053cce9a74311eeb6fc659d46c266e65dce28321/functions.php#L48

It seems like it depends on which LDAP server, it may not allow all user to bind. Currently, you are able to login with LDAP but there will be an error show after login. I am using the following to ignore error.
```php
if (@ldap_bind($ldap, sprintf(AUTHBACKENDDOMAINFORMAT, $username), $password))
```

 It is fine as incorrect password still cannot pass through.

I have looked up some LDAP implementation. It seems like the proper way is to use a admin LDAP account to bind the server and perform search. There also some just use `@` thought.

Also, the current LDAP use `sprintf` is not perfect. It need to prevent injection by escaping the user name. See [rfc4514](https://www.ietf.org/rfc/rfc4514.txt)
